stlab/concurrency/utility.hpp

```diff
@@ -137,104 +154,61 @@ T blocking_get(future<T> x) {
-            if (state->condition.wait_for(lock, timeout) == std::cv_status::timeout) {
-                hold.reset();
-                return stlab::optional<T>{};
-            }
-        }
+template <class T>
+struct blocking_get_guarded {
+    bool _timed_out{false};
+    future<T> _result;
+    std::condition_variable _condition;
+    std::mutex _mutex;
+
+    template <class F>
+    auto set(F&& r) -> future<T> {
+        std::unique_lock<std::mutex> lock{_mutex};
+        if (_timed_out) return std::forward<F>(r);
+        _result = std::forward<F>(r);
+        _condition.notify_one();
+        return make_exceptional_future<T>(
+            std::make_exception_ptr(std::runtime_error("not an error")), immediate_executor);
     }
 
-    if (state->error) std::rethrow_exception(state->error);
-
-    return std::move(state->result);
-}
-
-inline void blocking_get(future<void> x) {
-    std::exception_ptr error = nullptr;
-
-    bool flag{false};
-    std::condition_variable condition;
-    std::mutex m;
-    auto hold = std::move(x).recover(immediate_executor, [&](auto&& r) {
-        if (auto ex = std::forward<decltype(r)>(r).exception()) error = ex;
-        {
-            std::unique_lock<std::mutex> lock(m);
-            flag = true;
-            condition.notify_one();
-        }
-    });
-    {
-        std::unique_lock<std::mutex> lock(m);
-        while (!flag) {
-            condition.wait(lock);
-        }
+    auto wait_for(const std::chrono::nanoseconds& timeout) -> future<T> {
+        std::unique_lock<std::mutex> lock{_mutex};
+        _timed_out = !_condition.wait_for(lock, timeout, [&] { return _result.valid(); });
+        return _timed_out ? future<T>{} : std::move(_result);
     }
+};
 
-    if (error) std::rethrow_exception(error);
-}
+} // namespace detail
+
+template <class T>
+auto blocking_get_for(future<T> x, const std::chrono::nanoseconds& timeout) -> future<T> {
+    auto p = std::make_shared<detail::blocking_get_guarded<T>>();
 
-inline bool blocking_get(future<void> x, const std::chrono::nanoseconds& timeout) {
-    auto state = std::make_shared<detail::shared_state>();
-    auto hold =
-        std::move(x).recover(immediate_executor, [_weak_state = make_weak_ptr(state)](auto&& r) {
-            auto state = _weak_state.lock();
-            if (!state) {
-                return;
-            }
-            if (auto ex = r.exception()) state->error = ex;
-            {
-                std::unique_lock<std::mutex> lock(state->m);
-                state->flag = true;
-                state->condition.notify_one();
-            }
+    auto hold = std::move(x).recover(
+        immediate_executor, [_p = std::weak_ptr<detail::blocking_get_guarded<T>>(p)](auto&& r) {
+            if (auto p = _p.lock()) return p->set(std::forward<decltype(r)>(r));
+            return make_exceptional_future<T>(
+                std::make_exception_ptr(std::runtime_error("not an error")), immediate_executor);
```

Comment:
Here and above:

Isn't this exception a bit too unspecific? If the customer uses std::runtime_exeception as well, then the ambiguity is only specified by the error text.
I think introducing a new error code in future.hpp: 56 would be more concrete. Even that there are currently only error situations combined that can come from the future itself and not of a utility.

Reply:
The exception never escapes to the client. The problem here is that I need to return a future<T> and I can't create it with `make_ready_future` because that would require T is default constructible. I tried a default constructed future<T> but the downstream code fails badly with an future that isn't valid - so I return an exception.